mountpoint-s3-crt/src/auth/imds_client.rs

```diff
@@ -0,0 +1,127 @@
+        auth_library_init(allocator);
+
+        // SAFETY: `config` is moved into the [ImdsClient] on success, so `config.inner` (and the values
+        // inside) are guaranteed to live at least as long as this [ImdsClient] does. `allocator` is
+        // guaranteed to be a valid allocator because of the type-safe wrapper.
+        let inner = unsafe { aws_imds_client_new(allocator.inner.as_ptr(), &config.inner).ok_or_last_error()? };
+        Ok(Self { inner })
+    }
+
+    /// Get the ec2 instance type.
+    pub fn get_instance_type<F>(&self, callback: F) -> Result<(), Error>
+    where
+        F: FnOnce(Result<String, Error>) + 'static,
+    {
+        let callback_closure = Box::new(callback);
+        let callback_wrapper = Box::new(ImdsClientGetResourceCallback(Box::new(callback_closure)));
+        let callback_raw_ptr = Box::into_raw(callback_wrapper) as *mut libc::c_void;
+        let get_resource_callback_fn_ptr: Option<unsafe extern "C" fn(*const aws_byte_buf, i32, *mut libc::c_void)> =
+            Some(imds_client_get_resource_callback);
+
+        // SAFETY: `get_resource_callback_fn_ptr` is a guaranteed non-null callback function pointer pass to CRT.
+        // `callback_raw_ptr` is a pointer to a valid rust closure and the binding function will drop it once the
+        // corresponding crt function finish the execution.
+        // See `extern "C" fn imds_client_get_resource_callback`.
+        unsafe {
+            aws_imds_client_get_instance_type(self.inner.as_ptr(), get_resource_callback_fn_ptr, callback_raw_ptr)
+                .ok_or_last_error()
+        }
+    }
+}
+
+impl Drop for ImdsClient {
+    fn drop(&mut self) {
+        // SAFETY: `self.inner` always points to a valid underlying `aws_imds_client`.
+        // when `aws_imdw_client` creation failed, crt frees the allocated memroy
+        // otherwise return the pointer to `aws_imds_client`.
+        unsafe { aws_imds_client_release(self.inner.as_ptr()) };
+    }
+}
+
+/// Callback trait for when finished getting ec2 instance type.
+type OnGetResource = Box<dyn FnOnce(Result<String, Error>)>;
+struct ImdsClientGetResourceCallback(OnGetResource);
+
+/// Rust binding for CRT's callback function `aws_imds_client_on_get_resource_callback_fn`.
+unsafe extern "C" fn imds_client_get_resource_callback(
+    resource: *const aws_byte_buf,
+    error_code: i32,
+    user_data: *mut libc::c_void,
+) {
+    // SAFETY: CRT invokes the callback only once given it has `FnOnce` trait, the execution of callback
+    // rust closure is in unsafe zone, however, the borrow checker assures that instuctions
+    // will not result memory safety issue.
+    // Deferenece resource pointer only happenes when crt successful retrieves instance type, and since
+    // it is immutable pointer, no potential undefined behavior could occur to the actual data resource
+    // points to.
+    let callback = Box::from_raw(user_data as *mut ImdsClientGetResourceCallback).0;
```

Comment:
There's a few typos in here.

I'm also not very confident in the explanation of the safety, can we elaborate? (Not necessarily in code, we can discuss safety here and updated comment if necessary.)

I think lines 114-116 belong with the `*resource` de-referencing on line 122.

Reply:
I think think the comment needs to focus on explain why deference raw pointers are safe.
two  raw pointers `resource: *const aws_byte_buf` and `user_data: *mut libc::c_void`

1. `resource: *const aws_byte_buf`:  this raw pointer is the place where crt put the result bytes to, [if we crt won't return error code, then the instance type bytes are guaranteed in the memory it points to ](https://github.com/awslabs/aws-c-auth/blob/dd505b55fd46222834f35c6e54165d8cbebbfaaa/source/aws_imds_client.c#L610-L613) and thus deference  *const aws_byte_buf` is safe because a) it's a this raw pointer is imutable pointer and it won't now points to anywhere other than ec2 instance type bytes. b) when we deference this pointer, the pointer points to the valid data. 

2.  `user_data: *mut libc::c_void`: this is the actually rust closure passed to crt, acting as the C "function body". In rust, we define it type as `Box<dyn FnOnce(Result<String, Error>)>`  means whatever rust closure passed to CRT as callback function will be only execute once per request. Rust compiler assures the safety of closure. To be specific, we convert the closure outside the unsafe code block and we know the closure we wrapped a round is valid, hence within unsafe block, deference the raw pointer created outside the unsafe block is safe.